lib/src/multires_stochastic_texture_synthesis.rs

```diff
@@ -1191,6 +1201,213 @@ impl PrerenderedU8Function {
     }
 }
 
+struct TreeGrid {
+    grid_width: u32,
+    grid_height: u32,
+    offset_x: i32,
+    offset_y: i32,
+    chunk_size: u32,
+    rtrees: Vec<RwLock<RTree<[i32; 2]>>>,
+}
+
+// This is a grid of rtrees
+// The idea is that most pixels after the first couple steps will have their neighbors close by
+impl TreeGrid {
+    pub fn new(width: u32, height: u32, chunk_size: u32, offset_x: u32, offset_y: u32) -> TreeGrid {
+        let mut rtrees: Vec<RwLock<RTree<[i32; 2]>>> = Vec::new();
+        let grid_width = width / chunk_size + 1;
+        let grid_height = height / chunk_size + 1;
+        rtrees.resize_with((grid_width * grid_height) as usize, || {
+            RwLock::new(RTree::new())
+        });
+        TreeGrid {
+            rtrees,
+            grid_width,
+            grid_height,
+            offset_x: offset_x as i32,
+            offset_y: offset_y as i32,
+            chunk_size,
+        }
+    }
+
+    #[inline]
+    fn get_space(&self, x: u32, y: u32) -> usize {
+        (x * self.grid_height + y) as usize
+    }
+
+    pub fn force_insert(&self, x: i32, y: i32) {
+        let my_tree_index = self.get_space(
+            ((x + self.offset_x) as u32) / self.chunk_size,
+            ((y + self.offset_y) as u32) / self.chunk_size,
+        );
+        self.rtrees[my_tree_index].write().unwrap().insert([x, y]);
+    }
+
+    pub fn clone_into_new_tree_grid(&self, other: &TreeGrid) {
+        for tree in &self.rtrees {
+            for coord in tree.read().unwrap().iter() {
+                other.force_insert((*coord)[0], (*coord)[1]);
+            }
+        }
+    }
+
+    pub fn get_k_nearest_neighbors(
+        &self,
+        x: u32,
+        y: u32,
+        k: usize,
+        result: &mut Vec<SignedCoord2D>,
+    ) {
+        let offset_x = x as i32 + self.offset_x;
+        let offset_y = y as i32 + self.offset_y;
+
+        let chunk_x = offset_x / self.chunk_size as i32;
+        let chunk_y = offset_y / self.chunk_size as i32;
+
+        struct ChunkSearchInfo {
+            x: i32,
+            y: i32,
+            center: bool,
+            closest_point_on_boundary_x: i32,
+            closest_point_on_boundary_y: i32,
+        }
+
+        // Assume that all k nearest neighbors are in these cells
+        // it looks like we are rarely wrong once enough pixels are filled in
+        let places_to_look = vec![
```

Comment:
Looks like we can avoid a heap allocation by just removing the `vec!` characters :)

Reply:
Thanks good catch!